core/src/bytecode/ast/pattern/bindings.rs

```diff
@@ -0,0 +1,181 @@
+                evariant_pat.inject_bindings(bindings, path, parent_deco)
+            }
+            PatternData::Or(or_pat) => or_pat.inject_bindings(bindings, path, parent_deco),
+            // Wildcard and constant patterns don't bind any variable
+            PatternData::Wildcard | PatternData::Constant(_) => (),
+        }
+    }
+}
+
+impl<'ast> InjectBindings<'ast> for RecordPattern<'ast> {
+    fn inject_bindings(
+        &self,
+        bindings: &mut Vec<(Vec<LocIdent>, LocIdent, Field<'ast>)>,
+        path: Vec<LocIdent>,
+        parent_extra: Option<&Field<'ast>>,
+    ) {
+        for field_pat in self.patterns.iter() {
+            // Field patterns have their own annotation, so there's no need to propagate
+            // `parent_extra` any further
+            field_pat.inject_bindings(bindings, path.clone(), None);
+        }
+
+        if let TailPattern::Capture(rest) = self.tail {
+            // If a contract is attached to the whole record pattern in the enclosing pattern, the
+            // rest doesn't exactly match this contract: there are some fields missing. Still, it
+            // sounds more useful to keep the whole metadata - including documentation - for
+            // autocompletion and the like, even if it's an over-approximation.
+            bindings.push((path, rest, parent_extra.cloned().unwrap_or_default()));
+        }
+    }
+}
+
+impl<'ast> InjectBindings<'ast> for ArrayPattern<'ast> {
+    fn inject_bindings(
+        &self,
+        bindings: &mut Vec<(Vec<LocIdent>, LocIdent, Field<'ast>)>,
+        path: Vec<LocIdent>,
+        _parent_extra: Option<&Field<'ast>>,
+    ) {
+        for subpat in self.patterns.iter() {
+            // Array elements shouldn't inherit the annotation from their parent (for once, they
+            // are of a different type), so we reset `parent_extra` to `None`.
+            subpat.inject_bindings(bindings, path.clone(), None);
+        }
+
+        if let TailPattern::Capture(rest) = self.tail {
+            bindings.push((path, rest, Default::default()));
+        }
+    }
+}
+
+impl<'ast> InjectBindings<'ast> for FieldPattern<'ast> {
+    fn inject_bindings(
+        &self,
+        bindings: &mut Vec<(Vec<LocIdent>, LocIdent, Field<'ast>)>,
+        mut path: Vec<LocIdent>,
+        _parent_extra: Option<&Field<'ast>>,
+    ) {
+        path.push(self.matched_id);
+        self.pattern
+            .inject_bindings(bindings, path, Some(&Field::from(self.annotation.clone())));
+    }
+}
+
+impl<'ast> InjectBindings<'ast> for EnumPattern<'ast> {
+    fn inject_bindings(
+        &self,
+        bindings: &mut Vec<(Vec<LocIdent>, LocIdent, Field<'ast>)>,
+        path: Vec<LocIdent>,
+        _parent_extra: Option<&Field<'ast>>,
+    ) {
+        //TODO: I'm not sure we should just transparently forward to the variant's argument. Maybe
+        //we need a more complex notion of path here, that knows when we enter an enum variant?
```

Comment:
I think it isn't urgent. The two users of `bindings` are let pattern blocks (which ignores the path) and nls. Nls will ignore a path that passes through an enum.

Reply:
Yeah, this comment was probably just copy pasted from `term::pattern`.